sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,641 @@
+			checkNotNull(valueCoder, "ValueCoder should not be null in HadoopInputFormatSource");
+		}
+		boolean isSplitted=false;
+		@Override
+		public List<BoundedSource<KV<K, V>>> splitIntoBundles(long desiredBundleSizeBytes,PipelineOptions options) throws Exception {
+			// add comments
+			if (serializableSplit == null) {
+
+				if (serInputSplitList == null) {
+					computeSplits();
+				}
+				LOG.info("Generated {} splits each of size {} ",serInputSplitList.size(),serInputSplitList.get(0).getSplit().getLength());
+				return Lists.transform(serInputSplitList, new Function<SerializableSplit, BoundedSource<KV<K, V>>>() {
+					@Override
+					public BoundedSource<KV<K, V>> apply(SerializableSplit serializableInputSplit) {
+						HadoopInputFormatBoundedSource<K, V> hifBoundedSource = new HadoopInputFormatBoundedSource<K, V>(
+								conf, keyCoder, valueCoder, simpleFuncForKeyTranslation, simpleFuncForValueTranslation,
+								serializableInputSplit);
+						isSplitted=true;
+						return hifBoundedSource;
+					}
+				});
+			} else {
+				isSplitted=true;
+				LOG.info("Not splitting source {} because source is already split.", this);
+				return ImmutableList.of((BoundedSource<KV<K, V>>) this);
+			}
+
+		}
+
+		@VisibleForTesting
+		private void computeSplits() throws IOException, IllegalAccessException, InstantiationException,
+		InterruptedException, ClassNotFoundException {
+			Job job = Job.getInstance(conf.getConfiguration());
+			List<InputSplit> splits = job.getInputFormatClass().newInstance().getSplits(job);
+			if(splits == null){
+				LOG.error("Cannot split the source as getSplits() is returning null value.");
+				throw new IOException("Cannot split the source as getSplits() is returning null value.");
+			}
+			if(splits.isEmpty()){
+				LOG.error("Cannot split the source as getSplits() is returning empty list.");
+				throw new IOException("Cannot split the source as getSplits() is returning empty list.");
+			}
+			boundedSourceEstimatedSize=0;
+			serInputSplitList = new ArrayList<SerializableSplit>();
+			for (InputSplit inputSplit :splits) {
+				boundedSourceEstimatedSize = boundedSourceEstimatedSize + inputSplit.getLength();
+				serInputSplitList.add(new SerializableSplit(inputSplit));
+			}
+
+		}
+
+		List<SerializableSplit> serInputSplitList;
+		long boundedSourceEstimatedSize = 0;
+
+		@Override
+		public long getEstimatedSizeBytes(PipelineOptions po) throws Exception {
+			if (serInputSplitList == null) {
+				computeSplits();
+			}
+			return boundedSourceEstimatedSize;
+		}
+
+		@Override
+		public boolean producesSortedKeys(PipelineOptions po) throws Exception {
+			return false;
+		}
+
+		@Override
+		public Coder<KV<K, V>> getDefaultOutputCoder() {
+			return KvCoder.of(keyCoder, valueCoder);
+		}
+
+		@Override
+		public BoundedReader<KV<K, V>> createReader(PipelineOptions options) throws IOException {
+			this.validate();
+			if (serializableSplit == null) {
+				if(!isSplitted){
```

Comment:
@diptikul It seems like what you did was switch this to use isSourceSplit() - which doesn't seem very different. 

I still have the same question: What value does this add? You're adding a new variable (to a class that's supposed to be immutable), and the only value we're getting out of it is the differentiated error messages here. However, you'd have thrown that same error computeSplits().